runtime/v2/runc/task/service.go

```diff
@@ -113,7 +113,7 @@ type service struct {
 	containers map[string]*runc.Container
 
 	lifecycleMu sync.Mutex
-	running     map[int][]containerProcess // pid -> running process, guarded by lifecycleMu
+	running     map[int]map[containerProcess]struct{} // pid -> running process, guarded by lifecycleMu
```

Comment:
Commenting here to get some input: the fix for the failure mentioned in https://github.com/containerd/containerd/pull/8617#issuecomment-1581737795 has to do with us creating duplicate entries for the same `containerProcess` in `s.running` (first in `s.Create`, then in `s.Start`).

I fixed this in https://github.com/containerd/containerd/pull/8617/commits/c43966ced8998ffda1a99bafb81f7cfbc21fee7a by storing replacing the `containerProcess` slice with a map, and using the `containerProcess` as the key so that we don't keep the same `containerProcess` there twice. This works fine, but I wonder if we can make this more simple: I don't think we can assume `pid`s are unique if so, we could do:
```suggestion
	running     map[int]containerProcess // pid -> running process, guarded by lifecycleMu
```
If we can't assume that, I still wonder if we should do something like
```suggestion
	running     map[int]map[string]containerProcess // pid -> running process, guarded by lifecycleMu
```
where the key for the inner map is something like `container.ID`, since I think containerID+processID should be unique.

Reply:
@corhere thoughts?

Reply:
Silly question; if `pid` here is `containerProcess.Process.Pid`, could it then be `map[containerProcess]struct{}` ? (or `map[containerID]map[containerProcess]struct{}` if we need to index per container?

Reply:
Ah, ugh... gotcha

Reply:
@laurazard and I have concluded that the duplicate entries in `s.running` is really a symptom of an underlying bug in my implementation which could manifest as an exit event being published before a start event in certain cases. Deduplicating the `running` map values by `containerProcess` does not address the root cause. The correct solution, we believe, is to have `(*service).Start()` remove the container init process from `s.running` before calling `container.Start()` and add it back afterwards so that the process exiting before `(*service).Start()` publishes the `TaskStart` event is correctly handled as an early exit.